hathor/indexes/utxo_index.py

```diff
@@ -0,0 +1,326 @@
+            ))
+        # the spent inputs have now become unspent, re-add them to the index
+        for tx_input in tx.inputs:
+            spent_tx = tx.get_spent_tx(tx_input)
+            spent_tx_meta = spent_tx.get_metadata()
+            spent_tx_output = spent_tx.outputs[tx_input.index]
+            address_script = parse_address_script(spent_tx_output.script)
+            if not address_script:
+                # XXX: completely ignore scripts that aren't "addressable"
+                continue
+            if spent_tx.is_block:
+                # XXX: timelock forced to None when there is a heightlock
+                timelock = None
+                heightlock = spent_tx_meta.height + settings.REWARD_SPEND_MIN_BLOCKS
+            else:
+                timelock = address_script.get_timelock()
+                heightlock = None
+            assert timelock is None or heightlock is None
+            self._add_utxo(UtxoIndexItem(
+                token_uid=spent_tx.get_token_uid(spent_tx_output.get_token_index()),
+                tx_id=tx_input.tx_id,
+                index=tx_input.index,
+                address=address_script.address,
+                amount=spent_tx_output.value,
+                timelock=timelock,
+                heightlock=heightlock,
+            ))
+        self._remove_tx(tx.hash)
+
+    def iter_utxos(self, *, token_uid: bytes, address: str, target_amount: int, target_timestamp: Optional[int] = None,
+                   target_height: Optional[int] = None) -> Iterator[UtxoIndexItem]:
+        iter_nolock = self._iter_utxos_nolock(token_uid=token_uid, address=address,
+                                              target_amount=target_amount)
+        iter_timelock = self._iter_utxos_timelock(token_uid=token_uid, address=address,
+                                                  target_amount=target_amount,
+                                                  target_timestamp=target_timestamp)
+        iter_heightlock = self._iter_utxos_heightlock(token_uid=token_uid, address=address,
+                                                      target_amount=target_amount,
+                                                      target_height=target_height)
+        iter_utxos = peekable(sorted_merger(
+            iter_nolock,
+            iter_timelock,
+            iter_heightlock,
+            key=lambda item: (item.amount, item.tx_id, item.index),
+        ))
+
+        next_higher: Optional[UtxoIndexItem] = None
+        amount_sum = 0
+        count_utxos = 0
+        # we may have up to 3 items with amount higher than target_amount, skip until we get one lower, then yield it
+        # and start counting the amount_sum
+        for utxo_item in iter_utxos:
+            if utxo_item.amount >= target_amount:
+                next_higher = utxo_item
+            else:
+                amount_sum += utxo_item.amount
+                count_utxos += 1
+                if next_higher is not None:
```

Comment:
In your first example, why don't you return 7? If you can fulfill the requested amount with a single utxo, this is good, isn't it?

Reply:
I think this is part of the behavior we want. It could be more useful to use a range of smaller outputs instead of spending a larger or exact size one. If we only return 7, it's harder to choose.

Reply:
I'm a bit confused then.

I thought @pedroferreira1 meant that, when querying for 7, instead of returning 7, 5, 3, we would only return 7. Which I don't think it's what the API is meant for.

From what @msbrogli is saying, I'm now thinking that both of you mean that in the query for 5, instead of 5, 3, 1 it should return 7, 5. Is that it?

That's a simple change to make. The criteria I added was that the "extra UTXO" is one where `amount >= target_amount` and I continue walking the sorted list and start to count until `sum(value) >= target_amount`. The change would be for the "extra UTXO" to be one where `amount > target_amount`, if any.

I don't think there's any downside to doing that other than potentially ending up returning less UTXOs.

Reply:
It makes total sense now, I misunderstood the return of the method and the use of `next_higher`.